pallets/session-ext/src/impls.rs

```diff
@@ -0,0 +1,439 @@
+        validator_totals: &BTreeMap<T::AccountId, u32>,
+    ) -> impl IntoIterator<Item = (T::AccountId, u32)> + '_ {
+        committee.into_iter().map(move |validator| {
+            let total = BTreeMap::<_, _>::get(validator_totals, &validator).unwrap_or(&0);
+            let blocks_created = SessionValidatorBlockCount::<T>::get(&validator);
+            (
+                validator,
+                calculate_adjusted_session_points(
+                    nr_of_sessions,
+                    blocks_per_session,
+                    blocks_created,
+                    *total,
+                ),
+            )
+        })
+    }
+    fn blocks_to_produce_per_session() -> u32 {
+        T::SessionPeriod::get().saturating_div(
+            T::ValidatorProvider::current_era_committee_size()
+                .unwrap_or_default()
+                .size(),
+        )
+    }
+
+    pub fn adjust_rewards_for_session() {
+        if T::EraInfoProvider::active_era().unwrap_or(0) == 0 {
+            return;
+        }
+
+        let SessionValidators {
+            committee,
+            non_committee,
+        } = T::ValidatorProvider::current_session_committee_and_non_committee();
+        let nr_of_sessions = T::EraInfoProvider::sessions_per_era();
+        let blocks_per_session = Self::blocks_to_produce_per_session();
+        let validator_total_rewards = ValidatorEraTotalReward::<T>::get()
+            .unwrap_or_else(|| ValidatorTotalRewards(BTreeMap::new()))
+            .0;
+
+        let rewards = Self::reward_for_session_non_committee(
+            non_committee,
+            nr_of_sessions,
+            blocks_per_session,
+            &validator_total_rewards,
+        )
+        .into_iter()
+        .chain(
+            Self::reward_for_session_committee(
+                committee,
+                nr_of_sessions,
+                blocks_per_session,
+                &validator_total_rewards,
+            )
+            .into_iter(),
+        );
+
+        T::ValidatorRewardsHandler::add_rewards(rewards);
+    }
+
+    pub fn rotate_committee(current_session: SessionIndex) -> Option<Vec<T::AccountId>>
+    where
+        T::AccountId: Clone + PartialEq,
+    {
+        let EraValidators {
+            reserved,
+            non_reserved,
+        } = T::ValidatorProvider::current_era_validators()?;
+        let CommitteeSeats {
+            reserved_seats,
+            non_reserved_seats,
+        } = T::ValidatorProvider::current_era_committee_size()?;
+
+        rotate(
+            current_session,
+            reserved_seats as usize,
+            non_reserved_seats as usize,
+            reserved,
+            non_reserved,
+        )
+    }
+
+    pub(crate) fn calculate_underperforming_validators() {
+        let thresholds = BanConfig::<T>::get();
+        let SessionValidators {
+            committee: current_committee,
+            ..
+        } = T::ValidatorProvider::current_session_committee_and_non_committee();
```

Comment:
Couldn't you store this info during rotate? This would ultimately untangle the dependency and there would be no word `session` in elections pallet :)

Reply:
done